docs/DesignDocs/SGX_0-base_Enclave.md

```diff
@@ -0,0 +1,228 @@
+
+```C
+typedef struct oe_sgx_enclave_config_t
+{
+    uint16_t product_id;
+    uint16_t security_version;
+
+    /* Flag to indicate 0-base enclave, keeping alignment */
+    uint32_t flag_0_base;
+
+    uint8_t family_id[16];
+    uint8_t extended_product_id[16];
+    /* (OE_SGX_FLAGS_DEBUG | OE_SGX_FLAGS_MODE64BIT | OE_SGX_FLAGS_KSS) */
+    uint64_t attributes;
+
+    /* XSave Feature Request Mask */
+    uint64_t xfrm;
+
+    /* 0-base enclave start addr */
+    uint64_t start_addr;
+} oe_sgx_enclave_config_t;
+
+struct _oe_sgx_load_context
+{
+    oe_sgx_load_type_t type;
+    oe_sgx_load_state_t state;
+
+    /* attributes includes:
+     *  - OE_FLAG bits to be applied to the enclave, such as debug.
+     *  - XFRM supported by the OS to be used in enclave creation.
+     */
+    sgx_attributes_t attributes;
+
+    /* Fields used when attributes contain OE_FLAG_SIMULATION */
+    struct
+    {
+        /* Base address of enclave */
+        void* addr;
+
+        /* Size of enclave in bytes */
+        size_t size;
+    } sim;
+
+    /* Hash context used to measure enclave as it is loaded */
+    oe_sha256_context_t hash_context;
+
+#ifdef OE_WITH_EXPERIMENTAL_EEID
+    /* EEID data needed during enclave creation */
+    oe_eeid_t* eeid;
+#endif
+
+    const oe_config_data_t* config_data;
+    bool use_config_id;
+    /* If true, the enclave to be loaded is a 0-base enclave */
+    bool zero_base_enclave;
+    /* 0-base enclave start addr, valid only if zero_base_enclave is true  */
+    uint64_t start_addr;
+};
+```
+
+In `oe_sgx_create_enclave(...)` function, based on `zero_base_enclave` setting
+in the `oe_sgx_load_context` input, the function will invoke
+`libsgx_enclave_common.so` of the Intel SGX SW stack, which is considered part
+of the untrusted SGX Enclave Loader,  to create a regular SGX Enclave or a SGX
+Enclave with a fixed base of 0.
+
+When creating an SGX Enclave with a fixed base,  `libsgx_enclave_common.so`
+attempts to reserve the virtual memory at the requested `start_addr` and set
+`SECS.BASEADDR` as the requested value of 0. If the operations are successful,
+the returned enclave start address should match the requested `start_addr`.
+
+With `SECS.BASEADDR` as zero, and the enclave start address not matching
+`SECS.BASEADDR`, the enclave EPC page virtual address offset used in `EADD`
+instruction should be calculated based on `SECS.BASEADDR`. The offset
+calculation for `EADD` instruction is handled in the `libsgx_enclave_common.so`
+layer or the SGX driver layer. In the OE SDK, only
+`oe_sgx_measure_load_enclave_data(...)` call flow used during enclave signing
+needs to be adjusted to address the offset calculation. In the SGX `TCS` page,
+several fields are defined as offset value from `SECS.BASEADDR`. OE SDK's
+`_add_control_pages(...)` function needs to take into account that 0-base
+enclave's `SECS.BASEADDR` is 0, not `start_addr`, when filling in the SGX `TCS`
+page.
+
+## Authors
```

Comment:
The design proposal template suggests an Alternatives section, it might be worthwhile to add a paragraph just talking about other options in this space that were rejected (OS changes etc.)

Reply:
Added a "Alternative" section in the updated PR.